StyleCop.Analyzers/StyleCop.Analyzers/Helpers/CustomBatchFixAllProvider.cs

```diff
@@ -0,0 +1,558 @@
+﻿namespace StyleCop.Analyzers.Helpers
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CodeActions;
+    using Microsoft.CodeAnalysis.CodeFixes;
+    using Microsoft.CodeAnalysis.Text;
+
+    /// <summary>
+    /// Helper class for "Fix all occurrences" code fix providers.
+    /// </summary>
+    internal partial class CustomBatchFixAllProvider : FixAllProvider
+    {
+        protected CustomBatchFixAllProvider()
+        {
+        }
+
+        public static FixAllProvider Instance { get; } = new CustomBatchFixAllProvider();
+
+        public override async Task<CodeAction> GetFixAsync(FixAllContext fixAllContext)
+        {
+            if (fixAllContext.Document != null)
+            {
+                var documentsAndDiagnosticsToFixMap = await this.GetDocumentDiagnosticsToFixAsync(fixAllContext).ConfigureAwait(false);
+                return await this.GetFixAsync(documentsAndDiagnosticsToFixMap, fixAllContext).ConfigureAwait(false);
+            }
+            else
+            {
+                var projectsAndDiagnosticsToFixMap = await this.GetProjectDiagnosticsToFixAsync(fixAllContext).ConfigureAwait(false);
+                return await this.GetFixAsync(projectsAndDiagnosticsToFixMap, fixAllContext).ConfigureAwait(false);
+            }
+        }
+
+        public virtual async Task<CodeAction> GetFixAsync(
+            ImmutableDictionary<Document, ImmutableArray<Diagnostic>> documentsAndDiagnosticsToFixMap,
+            FixAllContext fixAllContext)
+        {
+            if (documentsAndDiagnosticsToFixMap != null && documentsAndDiagnosticsToFixMap.Any())
+            {
+                fixAllContext.CancellationToken.ThrowIfCancellationRequested();
+
+                var documents = documentsAndDiagnosticsToFixMap.Keys.ToImmutableArray();
+                var fixesBag = new List<CodeAction>[documents.Length];
+                var options = new ParallelOptions() { CancellationToken = fixAllContext.CancellationToken };
+                Parallel.ForEach(documents, options, (document, state, index) =>
+                {
+                    fixAllContext.CancellationToken.ThrowIfCancellationRequested();
+                    fixesBag[index] = new List<CodeAction>();
+                    this.AddDocumentFixesAsync(document, documentsAndDiagnosticsToFixMap[document], fixesBag[index].Add, fixAllContext).Wait(fixAllContext.CancellationToken);
+                });
+
+                if (fixesBag.Any(fixes => fixes?.Count > 0))
```

Comment:
Actually I believe the `Parallel.ForEach` call would throw an exception here if cancellation was requested (the only way `fixesBag` could still contain a null entry here).